ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,895 @@
+              compactionLog.markCycleComplete();
+            }
+            break;
+          default:
+            throw new IllegalStateException("Illegal compaction phase: " + phase);
+        }
+      }
+      endCompaction();
+    } catch (InterruptedException | IOException e) {
+      throw new StoreException("Exception during compaction", e, StoreErrorCodes.Unknown_Error);
+    } finally {
+      runningLatch.countDown();
+    }
+  }
+
+  /**
+   * If a compaction was in progress during a crash/shutdown, fixes the state so that the store is loaded correctly
+   * and compaction can resume smoothly on a call to {@link #resumeCompaction()}. Expected to be called before the
+   * {@link PersistentIndex} is instantiated in the {@link BlobStore}.
+   * @throws IOException if the {@link CompactionLog} could not be created or if commit or cleanup failed.
+   * @throws StoreException if the commit failed.
+   */
+  private void fixStateIfRequired() throws IOException, StoreException {
+    if (CompactionLog.isCompactionInProgress(dataDir.getAbsolutePath(), storeId)) {
+      compactionLog = new CompactionLog(dataDir.getAbsolutePath(), storeId, storeKeyFactory, time);
+      CompactionLog.Phase phase = compactionLog.getCompactionPhase();
+      switch (phase) {
+        case COMMIT:
+          commit(true);
+          compactionLog.markCleanupStart();
+          // fall through to CLEANUP
+        case CLEANUP:
+          cleanup(true);
+          compactionLog.markCycleComplete();
+          break;
+        case COPY:
+          recoveryStartToken = compactionLog.getSafeToken();
+          // fall through to the break.
+        case PREPARE:
+        case DONE:
+          break;
+        default:
+          throw new IllegalStateException("Unrecognized compaction phase: " + phase);
+      }
+      if (compactionLog.getCompactionPhase().equals(CompactionLog.Phase.DONE)) {
+        endCompaction();
+      }
+    }
+  }
+
+  /**
+   * Checks the sanity of the provided {@code details}
+   * @param details the {@link CompactionDetails} to use for compaction.
+   * @throws IllegalArgumentException if any of the provided segments doesn't exist in the log or if one or more offsets
+   * in the segments to compact are in the journal.
+   */
+  private void checkSanity(CompactionDetails details) {
+    List<String> segmentsUnderCompaction = details.getLogSegmentsUnderCompaction();
+    LogSegment lastSegment = null;
+    // all segments should be available
+    for (String segmentName : segmentsUnderCompaction) {
+      lastSegment = srcLog.getSegment(segmentName);
+      if (lastSegment == null) {
+        throw new IllegalArgumentException(segmentName + " does not exist in the log");
+      }
+    }
+    // last offset to compact should be outside the range of the journal
+    Offset lastSegmentEndOffset = new Offset(lastSegment.getName(), lastSegment.getEndOffset());
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
```

Comment:
No it cannot be - That would be an illegal state. I don't think a check is required